### PR TITLE
Add utility function to extract token id from address

### PR DESCRIPTION
**Pull request by polarker**

### CODE REVIEW

File: lib/utils.ts [hunk 1]
```diff
@@ -117,3 +117,17 @@ const groupOfP2mpkhAddress = (address: Uint8Array): number => {
...
 }
+
+export function tokenIdFromAddress(address: string): string {
+  const decoded = bs58.decode(address)
+
+  if (decoded.length == 0) throw new Error('Address string is empty')
+  const addressType = decoded[0]
+  const addressBody = decoded.slice(1)
+
+  if (addressType == 0x03) {
```
Comment (lf94) [inline lib/utils.ts:128]: Could you please make `0x03` a constant? Maybe create an enum if there are more than one possible type.

File: test/utils-test.ts [hunk 2]
```diff
@@ -113,4 +113,16 @@ describe('utils', function () {
       expect(utils.groupOfAddress('n2pYTzmA27tkp7UNFPhMJpjz3jr5vgessxqJ7kwomBMF')).toBe(2),
       expect(utils.groupOfAddress('tLf6hDfrUugmxZhKxGoZMpAUBt3NcZ2hrTspTCmZ6JdQ')).toBe(3)
   })
+
+  it('should calculate the group of addresses', () => {
```
Comment (lf94) [inline test/utils-test.ts:117]: Could you please add a test for a completely invalid address string?

File: lib/utils.ts [hunk 3]
```diff
@@ -117,3 +117,17 @@ const groupOfP2mpkhAddress = (address: Uint8Array): number => {
 const groupOfP2shAddress = (address: Uint8Array): number => {
   return groupOfAddressBytes(address)
 }
+
+export function tokenIdFromAddress(address: string): string {
```
Comment (lf94) [inline lib/utils.ts:121]: I noticed this function doesn't do any sanity check that it's a real address string. It is still possible to pass in base58 encoded which is not an address. Could you please add the check and the corresponding test?

Comment (polarker) [reply lib/utils.ts:121]: I guess it's fine here that I only check what is needed. If I check address format for every function that accept the address string, it's probably too much?

Comment (lf94) [reply lib/utils.ts:121]: Wherever this function is called it will need the address checked.

Comment (lf94) [reply lib/utils.ts:128]: Sure, that is one case.

Comment (polarker) [reply lib/utils.ts:128]: Let's say you got an address string, you validate it if the source is not trustworthy. After validation, you use the address to: extract token id, get address balance, get contract state etc. I don't see the sanity check is a part of this function.

Comment (lf94) [reply lib/utils.ts:128]: > Let's say you got an address string, you validate it if the source is not trustworthy. After validation, you use the address to: extract token id, get address balance, get contract state etc.

I understood you the first time - that is one case.

Comment (polarker) [reply lib/utils.ts:128]: Sorry, I don't get this. Maybe we need another task to add another function for address sanity checking?

Comment (lf94) [reply lib/utils.ts:128]: Yes, if such a `isValidAddress` is missing, let's open another PR then if you believe it's non-trivial (you would know best). I was under the assumption such a critical function already exists.

File: test/utils-test.ts (hunk 2 above)
Comment (lf94) [reply test/utils-test.ts:117]: Just waiting for the test and I'll approve.

File: lib/utils.ts (hunk 3 above)
Comment (polarker) [reply lib/utils.ts:128]: We mostly count on our APIs (scala code with strong type for everything) for validation for now, as: 1. I do not work that much for SDK; 2. We move fast without testing too much. Most of tests are added by @nop33.

I don't see I have time to polish SDK's details in the near future. I will probably still focus more on adding new features to SDK

File: test/utils-test.ts (hunk 2 above)
Comment (polarker) [reply test/utils-test.ts]: Just added. There are many other places using `base58.decode` 😅 I will let you guys to figure out how to fix them, or if it's a priority to fix

**Review by lf94 — APPROVED** [review]

**Pull request merged**